zebra-state/src/service/finalized_state/disk_format/transparent.rs

```diff
@@ -14,37 +14,57 @@ use zebra_chain::{
     block::Height,
     parameters::Network::*,
     serialization::{ZcashDeserializeInto, ZcashSerialize},
-    transaction,
     transparent::{self, Address::*},
 };
 
-use crate::service::finalized_state::disk_format::{block::HEIGHT_DISK_BYTES, FromDisk, IntoDisk};
+use crate::service::finalized_state::disk_format::{
+    block::{
+        TransactionIndex, TransactionLocation, HEIGHT_DISK_BYTES, TRANSACTION_LOCATION_DISK_BYTES,
+    },
+    expand_zero_be_bytes, truncate_zero_be_bytes, FromDisk, IntoDisk,
+};
 
 #[cfg(any(test, feature = "proptest-impl"))]
 use proptest_derive::Arbitrary;
 
+#[cfg(any(test, feature = "proptest-impl"))]
+mod arbitrary;
+
 /// Transparent balances are stored as an 8 byte integer on disk.
 pub const BALANCE_DISK_BYTES: usize = 8;
 
-/// Output transaction locations are stored as a 32 byte transaction hash on disk.
+/// [`OutputIndex`]es are stored as 3 bytes on disk.
 ///
-/// TODO: change to TransactionLocation to reduce database size and increases lookup performance (#3953)
-pub const OUTPUT_TX_HASH_DISK_BYTES: usize = 32;
+/// This reduces database size and increases lookup performance.
+pub const OUTPUT_INDEX_DISK_BYTES: usize = 3;
 
-/// [`OutputIndex`]es are stored as 4 bytes on disk.
+/// [`OutputLocation`]s are stored as a 3 byte height, 2 byte transaction index,
+/// and 3 byte output index on disk.
 ///
-/// TODO: change to 3 bytes to reduce database size and increases lookup performance (#3953)
-pub const OUTPUT_INDEX_DISK_BYTES: usize = 4;
+/// This reduces database size and increases lookup performance.
+pub const OUTPUT_LOCATION_DISK_BYTES: usize =
+    TRANSACTION_LOCATION_DISK_BYTES + OUTPUT_INDEX_DISK_BYTES;
 
 // Transparent types
 
-/// A transaction's index in its block.
-#[derive(Copy, Clone, Debug, Eq, PartialEq, Ord, PartialOrd, Serialize, Deserialize)]
-#[cfg_attr(any(test, feature = "proptest-impl"), derive(Arbitrary))]
+/// A transparent output's index in its transaction.
+#[derive(Copy, Clone, Debug, Eq, PartialEq, Ord, PartialOrd, Hash, Serialize, Deserialize)]
 pub struct OutputIndex(u32);
 
 impl OutputIndex {
-    /// Create a transparent output index from the native index integer type.
+    /// Create a transparent output index from the Zcash consensus integer type.
+    ///
+    /// `u32` is also the inner type.
+    pub fn from_index(output_index: u32) -> OutputIndex {
```

Comment:
Optional idea: should these `from_*` methods be `impl From<*> for OutputIndex`?

Reply:
I'd prefer them to have different names, so it's clear what kind of indexes the code is converting.